packages/flutter_tools/test/integration.shard/transition_test_utils.dart

```diff
@@ -40,20 +40,23 @@ abstract class Transition {
   /// The default value, null, leaves the logging state unaffected.
   final bool? logging;
 
+  /// Whether to check the line for containing a substring or an exact match.
```

Comment:
I'd expect this to be part of `pattern`, and `pattern` is a parameter in methods here and a field in the `Barrier` subclass. Why is the `contains` field here?

Reply:
The current code uses `Pattern pattern`, which means the pattern can be `String` or `RegExp`.

The `String` cannot encode an orthogonal concept (whether it should check for equality or contains)

Reply:
All parameters take `Pattern` not `String` so it's not orthogonal. But I do agree that upgrading a string to a pattern and doing the escaping is a bit iffy. However, I do think the field should live inside `class Barrier extends Transition` where the field `final Pattern pattern;` lives to have the symmetry of `pattern` and `contains`.

Reply:
"contains" vs "equality" is an orthogonal concept to whether to `Pattern` (aka `RegExp` or `String`):

One can
* *contains*: via `string.contains(RegExp(...))` and `string.contains(string)`
* *equality*: via  `string.firstMatch(RegExp(...))?.group(0) == string` and `string == other`

It's just that the existing code has these inconsistencies in and doesn't do it like this.

Reply:
Don't regexes include `^` and `$` to signal whether they are contains or equality? It would feel rather weird to pass in a regex with `^` and `$` and pass `contains: true` or a regex without `^` and `$` and pass `contains: false`. I guess I was expecting what you call `* equality` to be the implementation, and it isn't. And then we wouldn't need a contains because I could simply express that in the regex itself.

Anyway, my real comment is mostly about moving `final bool contains;` next to `final Pattern pattern;`, these two belong together, in the same class.

Reply:
> Don't regexes include ^ and $ to signal whether they are contains or equality? 

The meaning of those characters depend on how they are used. They can mean the start and end of a line or start and end of a string.

> Anyway, my real comment is mostly about moving final bool contains; next to final Pattern pattern;, these two belong together, in the same class.

:+1: If you prefer, I can duplicate the field in the subclasses instead and then pass it up in all the method calls